### PR TITLE
[metadata manager m8s] [Result Paginator] impl for result pagination

### PR DESCRIPTION
**Pull request by josephjacobmorris**
**What type of PR is this?**

> /kind new feature

**What this PR does / why we need it**:
This PR implements result pagination , sorted by backend name in ascending order.

**Test Report Added?**:
> /kind TESTED

**Test Report**:
1. offset =0, limit =0
image

2. offset = valid (present in result) , limit =0 (return 0 doc)
image

3. offset = invalid (greater than number of result items  in the document) , limit = valid
image

4. offset = valid, limit = valid
image

5. offset=empty, limit=empty
image
 
6. sort=asc
image

7. sort=desc
image

8. sort=desc1(invalid)
image

### CODE REVIEW

**Review by josephjacobmorris — COMMENTED** [review]
Self-review done

File: metadata/pkg/result-paginator/ResultPaginator.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package result_paginator
```
Comment (PravinRanjan10) [inline metadata/pkg/result-paginator/ResultPaginator.go:1]: Please avoid adding new file.
1. Please don't create new file inside pkg folder(especially for query related functions). Better create a folder and then push all file inside that folder

File: metadata/pkg/result-paginator/ResultPaginator.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+	if offset >= lengthOfUnPaginatedResult {
+		return []*model.MetaBackend{}
+	} else {
+		lastResultIndex := getLastResultIndex(limit, offset, lengthOfUnPaginatedResult)
+		return unPaginatedResult[offset:lastResultIndex]
+	}
+}
+
+func getLastResultIndex(limit int32, offset int32, lengthOfUnPaginatedResult int32) int {
```
Comment (PravinRanjan10) [inline metadata/pkg/result-paginator/ResultPaginator.go:19]: this function may not required.

File: metadata/pkg/service/service.go [hunk 3]
```diff
@@ -131,11 +132,13 @@ func (f *metadataService) ListMetadata(ctx context.Context, in *pb.ListMetadataR
 		return err
 	}
 
-	var backendMetaDatas []*pb.BackendMetadata
+	log.Debugln("the un-paginated result is:", unPaginatedResult)
```
Comment (PravinRanjan10) [inline metadata/pkg/service/service.go:135]: format it properly

**Review by PravinRanjan10 — CHANGES_REQUESTED** [review]

File: metadata/pkg/service/service.go (hunk 3 above)
Comment (josephjacobmorris) [reply metadata/pkg/service/service.go]: removed the line

File: metadata/pkg/validator/validator.go [hunk 4]
```diff
@@ -32,11 +32,28 @@ func isValidQuery(in *pb.ListMetadataRequest) (bool, error) {
...
+	okie, err = isSortParamValid(in.SortOrder)
+	if !okie {
+		return okie, err
+	}
 	// validate region
 	okie, err = isValidRegion(in.Region, in.Type)
 	return okie, err
 }
 
+func isSortParamValid(sortOrder string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply metadata/pkg/validator/validator.go:59]: done

File: metadata/pkg/result-paginator/ResultPaginator.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"github.com/opensds/multi-cloud/metadata/pkg/model"
+	"math"
+)
+
+func Paginate(unPaginatedResult []*model.MetaBackend, limit int32, offset int32) []*model.MetaBackend {
+	lengthOfUnPaginatedResult := int32(len(unPaginatedResult))
+
+	if offset >= lengthOfUnPaginatedResult {
```
[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply metadata/pkg/result-paginator/ResultPaginator.go]: done

File: metadata/pkg/validator/validator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply metadata/pkg/validator/validator.go:59]: removed

[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply metadata/pkg/validator/validator.go:59]: removed

File: metadata/pkg/result-paginator/ResultPaginator.go (hunk 2 above)
Comment (josephjacobmorris) [reply metadata/pkg/result-paginator/ResultPaginator.go]: done

File: metadata/pkg/result-paginator/ResultPaginator.go (hunk 1 above)
Comment (josephjacobmorris) [reply metadata/pkg/result-paginator/ResultPaginator.go]: Can't find a more appropriate class to move the function into

**Review by PravinRanjan10 — APPROVED** [review]

**Comment by PravinRanjan10** [comment]
LGTM

**Pull request merged**